packages/rum-core/src/domain/rumEventsCollection/action/trackActions.ts

```diff
@@ -65,142 +55,198 @@ export interface ActionContexts {
+        }
+      } else if (idleEvent.end < startClocks.timeStamp) {
+        // If the clock is looking weird, just discard the action
+        singleClickPotentialAction.discard()
       } else {
-        discard()
+        // Else complete the action at the end of the page activity
+        singleClickPotentialAction.complete(idleEvent.end)
       }
+      endClick()
     },
     AUTO_ACTION_MAX_DURATION
   )
+
   let viewCreatedSubscription: Subscription | undefined
-  if (!isExperimentalFeatureEnabled('frustration-signals')) {
-    // New views trigger the discard of the current pending Action
-    viewCreatedSubscription = lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, discard)
+  if (!state.collectFrustrations) {
+    // TODO: remove this in a future major version. To keep retrocompatibility, end the action on a
+    // new view is created.
+    viewCreatedSubscription = state.lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, endClick)
   }
 
-  function complete(endTime: TimeStamp) {
-    cleanup()
-    const eventCounts = eventCountsSubscription.eventCounts
-    lifeCycle.notify(LifeCycleEventType.AUTO_ACTION_COMPLETED, {
-      counts: {
-        errorCount: eventCounts.errorCount,
-        longTaskCount: eventCounts.longTaskCount,
-        resourceCount: eventCounts.resourceCount,
-      },
-      duration: elapsed(startClocks.timeStamp, endTime),
-      id,
-      name,
-      startClocks,
-      type,
-      event,
-    })
-    onCompleteCallback(endTime)
-  }
+  const stopSubscription = state.stopObservable.subscribe(endClick)
 
-  function discard() {
-    cleanup()
-    onDiscardCallback()
-  }
+  function endClick() {
+    singleClickPotentialAction.notifyIfComplete()
 
-  function cleanup() {
-    stopWaitingIdlePage()
-    eventCountsSubscription.stop()
+    // Cleanup any ongoing process
+    singleClickPotentialAction.discard()
     if (viewCreatedSubscription) {
       viewCreatedSubscription.unsubscribe()
     }
+    stopWaitingIdlePage()
+    stopSubscription.unsubscribe()
   }
+}
+
+function newPotentialAction(
+  { lifeCycle, history, collectFrustrations }: TrackActionsState,
+  base: Pick<AutoAction, 'startClocks' | 'event' | 'name' | 'type'>
+) {
+  const id = generateUUID()
+  const historyEntry = history.add(id, base.startClocks.relative)
+  const eventCountsSubscription = trackEventCounts(lifeCycle)
+  let finalState: { isDiscarded: false; endTime: TimeStamp } | { isDiscarded: true } | undefined
+  const frustrations = new Set<FrustrationType>()
 
   return {
-    discard,
-    id,
-    startClocks,
+    base,
+    frustrations,
+
+    complete: (endTime: TimeStamp) => {
+      if (finalState) {
+        return
+      }
+      finalState = { isDiscarded: false, endTime }
+      historyEntry.close(getRelativeTime(endTime))
+      eventCountsSubscription.stop()
+      if (eventCountsSubscription.eventCounts.errorCount > 0) {
+        frustrations.add(FrustrationType.ERROR)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏Here you flag with FrustrationType.ERROR inside PotentialAction while for FrustrationType.DEAD you do it outside.  Could be nice to do it in a single place.